HW1/P5/P5_bfs.py

```diff
@@ -0,0 +1,119 @@
+# import findspark
+# findspark.init()
+import pyspark
+sc = pyspark.SparkContext()
+N = 32
+
+def link_string_to_KV(s):
+        src, dests = s.split(': ')
+        dests = [int(to) for to in dests.split(' ')]
+        return (int(src), dests)
+
+
+def copartitioned(RDD1, RDD2):
+    "check if two RDDs are copartitioned"
+    return RDD1.partitioner == RDD2.partitioner
+
+def filterer(i):
+    def filt(x):
+        return x==itr_acc.value - 1
+    return filt
+
+def get_first_non_empty_cogroup(x2):
+    
+    for elem in x2:
+        elem = list(elem)
+        if len(elem) > 0:
+            return elem[0]
+
+def shortest_path(graph, nodeA, nodeB, N):
+    
+    graph = graph.partitionBy(N).cache()
+    nodes = sc.parallelize([(nodeA, 0)]).partitionBy(N)
+    node_destination = [nodeB]
+    # a = a.map(lambda x: (x[0], (list(x[1][0]))[0], (list(x[1][1]))))
+
+    total_char_acc = sc.accumulator(1)
+    it_acc = sc.accumulator(0)
+    nodes_4 = sc.parallelize([]).partitionBy(N)
+
+    while True:
+        
+        it_acc += 1
+        dist = it_acc.value
+                
+        assert copartitioned(nodes, graph)
+        
+        nodes_1 = nodes.filter(lambda x: filterer(x[1])).join(graph)
+        
+        assert copartitioned(nodes, nodes_1)
+        
+        nodes_2 = nodes_1.flatMap(lambda x: x[1][1]).distinct().map(lambda x: (x, dist))
+
+        nodes_3 = nodes_1.map(lambda x: (x[0],x[1][1],dist))
+        nodes_3 = nodes_3.map(lambda x: [(x[0], i, dist) for i in x[1]])
+                
+        nodes_4 = nodes_3.union(nodes_4)
+        
+        assert copartitioned(nodes_3, nodes_4)
+
+        nodes = nodes.cogroup(nodes_2).map(lambda x: (x[0], get_first_non_empty_cogroup(list(x[1])))).partitionBy(N).cache()
+
+        
+        if nodes_2.map(lambda x: x[0]).filter(lambda x: node_destination[0] == x).count() > 0:
+            print "Distance:", dist
```

Comment:
same as my last comment, you could put this directly in the while loop. It is a minor change but it makes the code a little cleaner.